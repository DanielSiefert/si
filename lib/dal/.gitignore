refactor.pl
test-refactor.pl
