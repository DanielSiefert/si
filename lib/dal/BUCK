rust_library(
    name = "dal",
    edition = "2021",
    deps = [
        "//lib/council-server:council-server",
        "//lib/nats-subscriber:nats-subscriber",
        "//lib/object-tree:object-tree",
        "//lib/si-data-nats:si-data-nats",
        "//lib/si-data-pg:si-data-pg",
        "//lib/si-pkg:si-pkg",
        "//lib/telemetry-rs:telemetry",
        "//lib/veritech-client:veritech-client",
        "//third-party/rust:async-recursion",
        "//third-party/rust:async-trait",
        "//third-party/rust:base64",
        "//third-party/rust:chrono",
        "//third-party/rust:derive_more",
        "//third-party/rust:diff",
        "//third-party/rust:dyn-clone",
        "//third-party/rust:futures",
        "//third-party/rust:hex",
        "//third-party/rust:iftree",
        "//third-party/rust:jwt-simple",
        "//third-party/rust:lazy_static",
        "//third-party/rust:once_cell",
        "//third-party/rust:paste",
        "//third-party/rust:petgraph",
        "//third-party/rust:postgres-types",
        "//third-party/rust:rand",
        "//third-party/rust:refinery",
        "//third-party/rust:regex",
        "//third-party/rust:serde",
        "//third-party/rust:serde-aux",
        "//third-party/rust:serde_json",
        "//third-party/rust:serde_with",
        "//third-party/rust:sodiumoxide",
        "//third-party/rust:strum",
        "//third-party/rust:thiserror",
        "//third-party/rust:tokio",
        "//third-party/rust:tokio-stream",
        "//third-party/rust:ulid",
        "//third-party/rust:url",
    ],
    srcs = glob([
        "src/**/*.rs",
        "src/builtins/func/**",
        "src/builtins/schema/data/**/*.json",
        "src/builtins/schema/definitions/**/*.json",
        "src/migrations/**/*.sql",
        "src/queries/**/*.sql",
    ]),
    env = {
        "CARGO_MANIFEST_DIR": ".",
    },
    visibility = ["PUBLIC"],
)
