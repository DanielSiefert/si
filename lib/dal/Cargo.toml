[package]
name = "dal"
version = "0.1.0"
authors = ["Adam Jacob <adam@systeminit.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
async-recursion = "1.0.0"
async-trait = "0.1.51"
base64 = "0.13.0"
chrono = { version = "0.4.19", features = ["serde"] }
color-eyre = { version = "0.6.1" }
derive_more = "0.99.16"
diff = "0.1"
dyn-clone = "1.0.6"
faktory-async = { git = "https://github.com/systeminit/faktory-async", branch = "main" }
futures = "0.3.13"
itertools = "0.10.3"
jwt-simple = "0.11.0"
lazy_static = "1.4.0"
names = { version = "0.14.0", default-features = false }
paste = "1.0"
postgres-types = { version = "0.2.2", features = ["derive"] }
rand = "0.8.3"
refinery = "0.8.4"
regex = "1.5.4"
serde = { version = "1.0.123", features = ["rc", "derive"] }
serde-aux = "4.0.0"
serde_json = { version = "1.0.64", features = ["preserve_order"] }
serde_with = "2.0.0"
si-data = { path = "../../lib/si-data", features = ["nats", "pg"] }
sodiumoxide = "0.2.6"
strum = "0.24.0"
strum_macros = "0.24.0"
telemetry = { path = "../../lib/telemetry-rs" }
tempfile = "3.2.0"
thiserror = "1.0.24"
tokio = { version = "1.2.0", features = ["full", "tracing"] }
tokio-stream = "0.1.3"
url = "2.2.2"
uuid = { version = "1.0.0", features = ["v4"] }
veritech-client = { path = "../../lib/veritech-client" }
# TODO(fnichol): server only needed for testing **but** currently used in `src/test.rs` which is prod code, argh!!!
# This can likely be flipped down to `dev-dependencies` once refactoring is complete
veritech-server = { path = "../../lib/veritech-server" }
which = "4.3.0"

[dev-dependencies]
pretty_assertions_sorted = "1.1.2"
si-test-macros = { path = "../../lib/si-test-macros" }
tracing = { version = "0.1" }
tracing-subscriber = { version = "0.3.11", default-features = false, features = ["ansi", "env-filter", "fmt"] }
