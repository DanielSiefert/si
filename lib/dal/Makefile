include ../../mk/base.mk
include ../../mk/rust.mk

.DEFAULT_GOAL := build
WATCH_PATHS := . \
	../../lib/si-data \
	../../lib/si-data-nats \
	../../lib/si-data-pg \
	../../lib/si-std \
	../../lib/si-test-macros \
	../../lib/telemetry-rs \
	../../lib/veritech-client
