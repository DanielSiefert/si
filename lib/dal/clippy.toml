disallowed-methods = [
    { path = "dal::test::helpers::commit_and_continue", reason = "should not be left in tests as this could infect other tests" }
]
