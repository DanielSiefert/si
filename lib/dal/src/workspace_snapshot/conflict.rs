use serde::{Deserialize, Serialize};

use crate::{workspace_snapshot::NodeInformation, EdgeWeightKindDiscriminants};

/// Describe the type of conflict between the given locations in a
/// workspace graph.
#[remain::sorted]
#[derive(Debug, Copy, Clone, PartialEq, Eq, Deserialize, Serialize)]
pub enum Conflict {
    ChildOrder {
        onto: NodeInformation,
        to_rebase: NodeInformation,
    },
    ExclusiveEdgeMismatch {
        source: NodeInformation,
        destination: NodeInformation,
        edge_kind: EdgeWeightKindDiscriminants,
    },
    ModifyRemovedItem(NodeInformation),
    NodeContent {
        onto: NodeInformation,
        to_rebase: NodeInformation,
    },
    RemoveModifiedItem {
        container: NodeInformation,
        removed_item: NodeInformation,
    },
}
