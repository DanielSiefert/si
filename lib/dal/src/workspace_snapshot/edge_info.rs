use petgraph::prelude::{EdgeIndex, NodeIndex};

use crate::{EdgeWeightKindDiscriminants, WorkspaceSnapshotGraphV1};

#[derive(Debug, Copy, Clone)]
pub struct EdgeInfo {
    pub source_node_index: NodeIndex,
    pub target_node_index: NodeIndex,
    pub edge_kind: EdgeWeightKindDiscriminants,
    pub edge_index: EdgeIndex,
}

impl EdgeInfo {
    pub fn simple_debug_string(&self, graph: &WorkspaceSnapshotGraphV1) -> String {
        let source = graph.graph().node_weight(self.source_node_index);
        let target = graph.graph().node_weight(self.target_node_index);
        let edge_weight = graph.graph().edge_weight(self.edge_index);

        format!(
            "{:?} -- {:?} --> {:?}",
            source.map(|s| s.id()),
            edge_weight.map(|e| e.kind()),
            target.map(|t| t.id())
        )
    }
}
