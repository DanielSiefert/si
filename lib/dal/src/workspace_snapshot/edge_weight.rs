//! Edges

use std::collections::HashSet;

use serde::{Deserialize, Serialize};
use si_events::VectorClockChangeSetId;
use strum::EnumDiscriminants;
use thiserror::Error;

use crate::workspace_snapshot::vector_clock::{VectorClock, VectorClockError, VectorClockId};

use super::vector_clock::HasVectorClocks;

pub mod deprecated;

#[derive(Debug, Error)]
pub enum EdgeWeightError {
    #[error("Vector Clock error: {0}")]
    VectorClock(#[from] VectorClockError),
}

pub type EdgeWeightResult<T> = Result<T, EdgeWeightError>;

#[remain::sorted]
#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq, Hash, EnumDiscriminants)]
#[strum_discriminants(derive(Hash, Serialize, Deserialize))]
pub enum EdgeWeightKind {
    Action,
    /// A function used by a [`SchemaVariant`] to perform an action that affects its resource
    ActionPrototype,
    /// A function defined for a secret defining [`SchemaVariant`] to be executed before funcs on
    /// components that have a secret of that kind
    AuthenticationPrototype,
    /// An [`AttributeValue`] "contained" by another [`AttributeValue`], such as an entry in an
    /// array/map, or a field of an object. The optional [`String`] represents the key of the entry
    /// in a map.
    Contain(Option<String>),
    /// Used to indicate parentage within frames. It does not dictate data flow. That is provided via
    /// [`ComponentType`](crate::ComponentType).
    ///
    /// This replaces "Symbolic" edges and "Frame" sockets from the old engine.
    FrameContains,
    /// Used to record the order that the elements of a container should be presented in.
    Ordering,
    /// Connects the node at the Ordering edge directly to the things it orders.
    Ordinal,
    /// Used to link an attribute value to the prop that it is for.
    Prop,
    /// An edge from a [`socket`](crate::socket), [`AttributeValue`](crate::AttributeValue) or [`Prop`](crate::Prop)
    /// to an [`AttributePrototype`](crate::AttributePrototype). The optional [`String`] is used for
    /// maps, arrays and relevant container types to indicate which element the prototype is for.
    Prototype(Option<String>),
    /// An edge from an [`AttributePrototype`][crate::AttributePrototype] to an
    /// [`AttributePrototypeArgument`][crate::AttributePrototypeArgument].
    PrototypeArgument,
    /// An edge from an
    /// [`AttributePrototypeArgument`][crate::AttributePrototypeArgument] to the
    /// source for the value for this argument
    PrototypeArgumentValue,
    Proxy,
    /// Indicates the "root" [`AttributeValue`](crate::AttributeValue) for a [`Component`](crate::Component).
    ///
    /// TODO(nick): in the future, this should be used for the "root" [`Prop`](crate::Prop) for a
    /// [`SchemaVariant`](crate::SchemaVariant) as well.
    Root,
    /// Used when the target/destination of an edge is an [`InputSocket`](crate::InputSocket), or an
    /// [`OutputSocket`](crate::OutputSocket).
    Socket,
    /// Edge from component to input or output Socket's attribute value
    SocketValue,
    /// Workspaces "use" functions, modules, schemas. Schemas "use" schema variants.
    /// Schema variants "use" props. Props "use" functions, and other props. Modules
    /// "use" functions, schemas, and eventually(?) components.
    Use {
        is_default: bool,
    },
    /// Edge from attribute value to validation result node
    ValidationOutput,
}

impl EdgeWeightKind {
    /// Creates a new Use edge weight kind indicating that this is also the
    /// "default" use edge
    pub fn new_use_default() -> Self {
        EdgeWeightKind::Use { is_default: true }
    }

    /// Creates a non-default use EdgeWeightKind. This is what you normally want
    /// unless you know there should be a default/non-default difference
    pub fn new_use() -> Self {
        EdgeWeightKind::Use { is_default: false }
    }
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
pub struct EdgeWeight {
    kind: EdgeWeightKind,
    vector_clock_first_seen: VectorClock,
    vector_clock_recently_seen: VectorClock,
    vector_clock_write: VectorClock,
}

impl EdgeWeight {
    pub fn kind(&self) -> &EdgeWeightKind {
        &self.kind
    }

    /// Remove stale vector clock entries. `allow_list` should always include
    /// the current editing clock id...
    pub fn collapse_vector_clock_entries(
        &mut self,
        allow_list: &HashSet<VectorClockChangeSetId>,
        collapse_id: VectorClockId,
    ) {
        self.vector_clock_first_seen
            .collapse_entries(allow_list, collapse_id);
        self.vector_clock_recently_seen
            .collapse_entries(allow_list, collapse_id);
        self.vector_clock_write
            .collapse_entries(allow_list, collapse_id);
    }

    pub fn new(vector_clock_id: VectorClockId, kind: EdgeWeightKind) -> EdgeWeightResult<Self> {
        let empty_vector_clock = VectorClock::new(vector_clock_id);

        Ok(Self {
            kind,
            vector_clock_first_seen: empty_vector_clock.clone(),
            vector_clock_recently_seen: empty_vector_clock.clone(),
            vector_clock_write: empty_vector_clock,
        })
    }
}

impl HasVectorClocks for EdgeWeight {
    fn vector_clock_recently_seen(&self) -> &VectorClock {
        &self.vector_clock_recently_seen
    }

    fn vector_clock_recently_seen_mut(&mut self) -> &mut VectorClock {
        &mut self.vector_clock_recently_seen
    }

    fn vector_clock_first_seen(&self) -> &VectorClock {
        &self.vector_clock_first_seen
    }

    fn vector_clock_first_seen_mut(&mut self) -> &mut VectorClock {
        &mut self.vector_clock_first_seen
    }

    fn vector_clock_write(&self) -> &VectorClock {
        &self.vector_clock_write
    }

    fn vector_clock_write_mut(&mut self) -> &mut VectorClock {
        &mut self.vector_clock_write
    }
}
