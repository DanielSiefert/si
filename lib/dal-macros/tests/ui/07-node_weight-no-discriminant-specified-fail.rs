use dal::workspace_snapshot::{
    edge_weight::EdgeWeightKindDiscriminants,
    node_weight::{
        traits::{CorrectExclusiveOutgoingEdge, CorrectTransforms, SiNodeWeight},
        NodeWeightDiscriminants,
    },
};
use si_events::{merkle_tree_hash::MerkleTreeHash, ulid::Ulid, ContentHash};

#[derive(dal_macros::SiNodeWeight)]
#[si_node_weight]
pub struct TestingNodeWeight {
    id: Ulid,
    lineage_id: Ulid,
    merkle_tree_hash: MerkleTreeHash,
}

impl CorrectTransforms for TestingNodeWeight {}
impl CorrectExclusiveOutgoingEdge for TestingNodeWeight {
    fn exclusive_outgoing_edges(&self) -> &[EdgeWeightKindDiscriminants] {
        &[]
    }
}

fn main() {}
