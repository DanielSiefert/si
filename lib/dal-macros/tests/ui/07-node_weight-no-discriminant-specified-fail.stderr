error: No NodeWeightDiscriminants was specified.
  --> tests/ui/07-node_weight-no-discriminant-specified-fail.rs:11:1
   |
11 | / #[si_node_weight]
12 | | pub struct TestingNodeWeight {
13 | |     id: Ulid,
14 | |     lineage_id: Ulid,
15 | |     merkle_tree_hash: MerkleTreeHash,
16 | | }
   | |_^

error: unexpected end of input, expected identifier
  --> tests/ui/07-node_weight-no-discriminant-specified-fail.rs:10:10
   |
10 | #[derive(dal_macros::SiNodeWeight)]
   |          ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the derive macro `dal_macros::SiNodeWeight` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: unused imports: `NodeWeightDiscriminants`, `SiNodeWeight`
 --> tests/ui/07-node_weight-no-discriminant-specified-fail.rs:4:67
  |
4 |         traits::{CorrectExclusiveOutgoingEdge, CorrectTransforms, SiNodeWeight},
  |                                                                   ^^^^^^^^^^^^
5 |         NodeWeightDiscriminants,
  |         ^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: unused import: `ContentHash`
 --> tests/ui/07-node_weight-no-discriminant-specified-fail.rs:8:63
  |
8 | use si_events::{merkle_tree_hash::MerkleTreeHash, ulid::Ulid, ContentHash};
  |                                                               ^^^^^^^^^^^

error[E0277]: the trait bound `TestingNodeWeight: SiVersionedNodeWeight` is not satisfied
  --> tests/ui/07-node_weight-no-discriminant-specified-fail.rs:19:6
   |
19 | impl CorrectExclusiveOutgoingEdge for TestingNodeWeight {
   |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `SiVersionedNodeWeight` is not implemented for `TestingNodeWeight`, which is required by `for<'a> NodeInformation: From<&'a TestingNodeWeight>`
   |
   = help: the following other types implement trait `SiVersionedNodeWeight`:
             ManagementPrototypeNodeWeight
             dal::workspace_snapshot::node_weight::InputSocketNodeWeight
             dal::workspace_snapshot::node_weight::SchemaVariantNodeWeight
   = note: required for `TestingNodeWeight` to implement `SiNodeWeight`
   = note: required for `NodeInformation` to implement `for<'a> From<&'a TestingNodeWeight>`
note: required by a bound in `CorrectExclusiveOutgoingEdge`
  --> $WORKSPACE/lib/dal/src/workspace_snapshot/node_weight/traits/correct_exclusive_outgoing_edge.rs
   |
   | pub trait CorrectExclusiveOutgoingEdge
   |           ---------------------------- required by a bound in this trait
   | where
   |     NodeInformation: for<'a> From<&'a Self>,
   |                      ^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `CorrectExclusiveOutgoingEdge`
